src/hugr/hugrmut.rs

```diff
@@ -8,277 +8,413 @@ use portgraph::{LinkMut, NodeIndex, PortMut, PortView, SecondaryMap};
 use crate::hugr::{Direction, HugrError, HugrView, Node, NodeType};
 use crate::ops::OpType;
 
-use crate::ops::handle::NodeHandle;
 use crate::{Hugr, Port};
 
+use self::sealed::HugrMutInternals;
+
 use super::NodeMetadata;
 
-pub(crate) use sealed::HugrInternalsMut;
+/// Functions for low-level building of a HUGR.
+pub trait HugrMut: HugrView + HugrMutInternals {
+    /// Returns the metadata associated with a node.
+    fn get_metadata_mut(&mut self, node: Node) -> &mut NodeMetadata;
+
+    /// Sets the metadata associated with a node.
+    fn set_metadata(&mut self, node: Node, metadata: NodeMetadata) {
+        *self.get_metadata_mut(node) = metadata;
+    }
 
-pub trait HugrMut: sealed::HugrInternalsMut {
-    /// The kind of handle that can be used to refer to the root node.
+    /// Add a node to the graph with a parent in the hierarchy.
     ///
-    /// The handle is guaranteed to be able to contain the operation returned by
-    /// [`HugrView::root_type`].
-    type RootHandle: NodeHandle;
+    /// The node becomes the parent's last child.
+    #[inline]
+    fn add_op_with_parent(
+        &mut self,
+        parent: Node,
+        op: impl Into<OpType>,
+    ) -> Result<Node, HugrError> {
+        self.valid_node(parent)?;
+        self.hugr_mut().add_op_with_parent(parent, op)
+    }
 
-    /// The view interface for this mutable reference to a Hugr.
-    type View<'h>: 'h + HugrView<RootHandle = Self::RootHandle>
-    where
-        Self: 'h;
+    /// Add a node to the graph with a parent in the hierarchy.
+    ///
+    /// The node becomes the parent's last child.
+    #[inline]
+    fn add_node_with_parent(&mut self, parent: Node, op: NodeType) -> Result<Node, HugrError> {
+        self.valid_node(parent)?;
+        self.hugr_mut().add_node_with_parent(parent, op)
+    }
+
+    /// Add a node to the graph as the previous sibling of another node.
+    ///
+    /// The sibling node's parent becomes the new node's parent.
+    ///
+    /// # Errors
+    ///
+    ///  - If the sibling node does not have a parent.
+    ///  - If the attachment would introduce a cycle.
+    #[inline]
+    fn add_op_before(&mut self, sibling: Node, op: impl Into<OpType>) -> Result<Node, HugrError> {
+        self.valid_node(sibling)?;
```

Comment:
Ok, so the question here is - if "sibling" was the root (parent) of a SiblingGraph, would that be accepted by "valid_node"? By most definitions, that lets you add something outside the region, so we might need *two* versions of `valid_node`. (I *think* it's reasonable to add something as a child of one of the siblings in a SiblingGraph, even though the something-added would then not strictly be in the SiblingGraph...)

Reply:
Good catch!